tern/analyze/docker/container.py

```diff
@@ -192,3 +198,139 @@ def close_client():
+
+    return temp_path
+
+
+def generate_manifest_url(image_name, image_tag):
+    '''
+    Generate a docker image manifest url end point with
+    given image name and tag
+    '''
+    url = "{0}/{1}/manifests/{2}"
+    return url.format(docker_endpoint, image_name, image_tag)
+
+
+def generage_image_blobs_url(image_name, digest):
+    '''
+    Generate a docker image blobs url end point with
+    given image name and tag
+    '''
+    url = "{0}/{1}/blobs/{2}"
+    return url.format(docker_endpoint, image_name, digest)
+
+
+def is_token_expired():
+    '''method to validtae if Docker HTTP Rest API auth token is expired'''
+    global auth_token
+    token_expires_in = int(auth_token.get("expires_in"))
+    token_issued_at = auth_token.get("issued_at")
+    start = datetime.timestamp(parser.isoparse(token_issued_at))
+    end = datetime.timestamp(datetime.now())
+
+    if end - start > token_expires_in:
+        return True
+
+    return False
+
+
+def generate_auth_token(image_name, image_tag):
+    '''
+    method to generate Docker HTTP Rest API auth token
+    for given image_name and image_tag
+    '''
+    global auth_token
+    url = generate_manifest_url(image_name, image_tag)
+    response = requests.get(url)
+    authentication = response.headers.get("Www-Authenticate").split(",")
+    auth_url, params = None, dict()
+    for line in authentication:
+        if "Bearer realm" in line:
+            auth_url = json.loads(line.split("=")[1])
+        if "service" in line:
+            params["service"] = json.loads(line.split("=")[1])
+        if "scope" in line:
+            params["scope"] = json.loads(line.split("=")[1])
+    response = requests.get(auth_url, params=params)
+    if response.status_code != 200:
+        raise Exception(response.text)
+
+    auth_token = response.json()
+
+
+def pull_image_manifest(image_name, image_tag):
+    '''download and return the image manifest'''
+    global auth_token
+    url = generate_manifest_url(image_name, image_tag)
+    generate_auth_token(image_name, image_tag)
+    token_value = auth_token.get("token")
+    headers = dict()
+    headers['Authorization'] = 'Bearer {0}'.format(token_value)
+    headers['Accept'] = 'application/vnd.docker.distribution.manifest.v2+json'
+    response = requests.get(url, headers=headers)
+    if response.status_code != 200:
+        raise Exception(response.text)
+
+    return response.json()
+
+
+def pull_image_layers(image_name, image_tag, layers):
+    '''Given manifest data, saves the image layers inside mount dir'''
+    global auth_token
+    headers = dict()
+    headers['Accept'] = 'application/vnd.docker.image.rootfs.diff.tar.gzip'
+    for layer in layers:
+        digest = layer.get("digest")
+        url = generage_image_blobs_url(image_name, digest)
+        # to generate token if it is expired while downloading a layer
+        if is_token_expired():
+            generate_auth_token(image_name, image_tag)
```

Comment:
I believe you can get a refresh token so you don't have to keep checking if a token is expired.